<?xml version="1.0" encoding="UTF-8"?>
<launch>
    <arg name="use_rviz" default="False" />
    <arg name="use_camera" default="False" />

    <!-- Adjust the camera parameters to match your setup -->
    <arg name="camera_topic" default="/gazebo_camera/image_raw" />
    <arg name="camera_info_topic" default="/gazebo_camera/camera_info" />

    <!-- Camera node configuration -->
    <group if="$(arg use_camera)" >
        <node name="your_camera_node" pkg="your_camera_pkg" type="your_camera_node_type" output="screen">
            <param name="video_device" value="/dev/video1" />
            <param name="image_width" value="640" />
            <param name="image_height" value="480" />
            <param name="pixel_format" value="yuyv" />
            <param name="camera_frame_id" value="camera" />
            <param name="io_method" value="mmap"/>
        </node>
    </group>

    <!-- Camera info publication -->
    <node name="camera_info" pkg="rostopic" type="rostopic"
        args="pub $(arg camera_info_topic) sensor_msgs/CameraInfo
        '{header: {seq: 0, stamp: {secs: 0, nsecs: 0}, frame_id: 'camera1'},
        height: 480, width: 640, distortion_model: 'plumb_bob',
        D: [0.08269409104660361, -0.16570658726478218, 0.0013173694798400222, -0.0014110498528553663, 0.0],
        K: [519.734208757597, 0.0, 302.85096428096654, 0.0, 520.5923676830233, 252.63525359615707, 0.0, 0.0, 1.0],
        R: [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0],
        P: [523.3435668945312, 0.0, 301.43917490757667, 0.0, 0.0, 526.1812744140625, 253.1422791119403, 0.0, 0.0, 0.0, 1.0, 0.0],
        binning_x: 0, binning_y: 0,
        roi: {x_offset: 0, y_offset: 0, height: 480, width: 640, do_rectify: false}}' -r 2"/>

    <!-- ArUco marker detection node configuration -->
    <arg name="camera" default="/gazebo_camera/image_raw" />
    <arg name="camera_info" default="/gazebo_camera/camera_info" />
    <arg name="tf_prefix" default="marker_id" />
    <arg name="show_detections" default="true" />
    <arg name="marker_size" default="0.1" doc="length of the physical marker's side in meters"/>
    <arg name="dictionary_name" default="DICT_6X6_250" />
    <arg name="blur_window_size" default="7" />
    <arg name="num_detected" default="25" /><!-- 0: do not use it, -->
    <arg name="min_prec_value" default="80" /><!-- 0: do not use it, 80 means aruco marker must be detected in 80% of the last pics to be displayed -->
    <node pkg="aruco_detector_ocv" type="node" name="aruco_detector_ocv" output="screen">
        <param name="camera" type="string" value="$(arg camera)"/>
        <param name="camera_info" type="string" value="$(arg camera_info)"/>
        <param name="tf_prefix" type="string" value="$(arg tf_prefix)"/>
        <param name="show_detections" type="bool" value="$(arg show_detections)"/>
        <param name="marker_size" type="double" value="$(arg marker_size)"/>
        <param name="dictionary_name" type="string" value="$(arg dictionary_name)"/>
        <param name="image_width" type="double" value="640"/>
        <param name="image_height" type="double" value="480"/>
        <param name="blur_window_size" type="double" value="$(arg blur_window_size)"/>
        <param name="num_detected" type="double" value="$(arg num_detected)"/>
        <param name="min_prec_value" type="double" value="$(arg min_prec_value)"/>
    </node>

    <!-- Uncomment this section if you want to use RViz for visualization -->
    <!-- <group if="$(arg use_rviz)" >
        <node pkg="rviz" type="rviz" name="markers_extrinsic" args="-d $(find aruco_detector_ocv)/config/aruco_opencv.rviz"/>
    </group> -->
</launch>
